[Q3] Auto-update asset list at regular intervals

deividaspetraitis opened the pull request:
This PR implements updating asset list at regular intervals based on the procesing block in the ingester, please see #279 for more information.

## Implementation details

- As a side effect several more mocks were created in order to be able to test ingester use case, and existing mock was modified to support future testing requirements.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced auto-updating asset list at regular intervals to enhance system functionality and data accuracy.
  - Added a new configuration parameter to define the block interval at which assets are updated.

- **Documentation**
  - Updated README with a corrected description for a token metadata endpoint and information on the new block interval configuration parameter.

- **Bug Fixes**
  - Corrected naming inconsistencies and reorganized parameter calls in various functions for clarity and improved functionality.

- **Tests**
  - Added new test functions to ensure assets are updated at specified intervals and to validate the new functionalities introduced.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

deividaspetraitis committed:
[Q3] Auto-update asset list at regular intervals

deividaspetraitis commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -357,6 +363,64 @@ func (t *tokensUseCase) getChainScalingFactorMut(precision int) (osmomath.Dec, b
+)
+
+// StartAsRoutineIfNotRunning starts the job in a new goroutine if it is not already running.
+func StartAsRoutineIfNotRunning(job func()) bool {
```
Those functions are public, because tests resides in other package, making original package to leak implementation details. Should I just create different file that would be within same package for example `tokens_usecase_i_test.go` where I could place tests?

deividaspetraitis replied:
Removed this code, no longer relevant.

deividaspetraitis commented on tokens/usecase/tokens_usecase_test.go:
```diff
@@ -343,3 +345,69 @@ func (s *TokensUseCaseTestSuite) TestGetPrices_Chain_PricingOptions() {
+
+	var got int
+	for i := gen; i < gen+iterations+1; i++ {
+		tokensusecase.ExecuteOnTickInterval(uint64(i), every, func() {
```
See https://github.com/osmosis-labs/sqs/pull/323/files#r1640012014

deividaspetraitis committed:
Lint: fixes

deividaspetraitis commented:
hey @p0mvn, while there are still several things left to implement in this PR, can I ask you for your early feedback at this point to make sure I am on the right track and consistent with the project code base ( I have left a few clarifying questions above )?

deividaspetraitis commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -357,6 +363,64 @@ func (t *tokensUseCase) getChainScalingFactorMut(precision int) (osmomath.Dec, b
+}
+
+// UpdateAssetsAtHeightInterval updates assets at specified height intervals by running fn.
+// Function fn will be launched as separate goroutine.
```
TODO: add docs; if `fn` ( `LoadTokensFromChainRegistryFunc` ) is already executing it will not be called twice.

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -110,18 +122,12 @@ func NewTokensUsecase(tokenMetadataByChainDenom map[string]domain.Token) mvc.Tok
+	t.tokenMetadataByChainDenom = tokenMetadataByChainDenom
+	t.humanToChainDenomMap = humanToChainDenomMap
+	t.precisionScalingFactorMap = precisionScalingFactors
 
-		chainDenoms:  chainDenoms,
-		coingeckoIds: coingeckoIds,
-	}
+	t.chainDenoms = chainDenoms
+	t.coingeckoIds = coingeckoIds
```
I think we should make all of these `sync.Map`. Then, iterate over reach updated map and update it element-by-element.

In the current approach, we would get a concurrent map read and write.

In the alternative approach of surrounding these with locks, I'm afraid that we would get a tail latency increase due to copying entire maps under locks.

deividaspetraitis replied:
Thanks, I have updated code to reflect your idea. :)

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -357,6 +363,64 @@ func (t *tokensUseCase) getChainScalingFactorMut(precision int) (osmomath.Dec, b
+
+// StartAsRoutineIfNotRunning starts the job in a new goroutine if it is not already running.
+func StartAsRoutineIfNotRunning(job func()) bool {
+	jobIsrunningMu.Lock()
+	if jobIsRunning {
+		jobIsrunningMu.Unlock()
+		return false
+	}
+	jobIsRunning = true
+	jobIsrunningMu.Unlock()
+
+	go func() {
+		jobIsrunningMu.Lock()
+		job()
+		jobIsRunning = false
+		jobIsrunningMu.Unlock()
+	}()
```
Note that we don't really need this synchronization. We should be able to simply spin up a goroutine without any locks

deividaspetraitis replied:
Implementation reflects the idea that only one assets update go routine can be running at any given time. For example if routine is already running second routine would be not launched to avoid DOS'ing external service.

I thought it can be useful for some edge cases:

- Wrong configuration of the update frequency ( for example every height )
- Network latency, some routine may be still in progress while other is started
- TCP/IP connections ( sockets ) are files, unlimited number of open descriptor might lead to problems

Please let me know would you would like still to proceed to remove it and I will address that.

p0mvn replied:
I see. I would like to recommend considering using a worker pool [design pattern](https://shopify.engineering/leveraging-go-worker-pools) to address these edge cases.

We have the barebones abstractions defined [here](https://github.com/osmosis-labs/sqs/blob/v25.x/domain/workerpool/worker_pool.go)

By leveraging the abstraction, we will reuse the existing components, reducing complexity and addressing the raised concerns by configuring a pool with 1 concurrent worker.

However, I think that we can derisk them with a wider height interval that we can control with tests and observability.

I'll leave the worker pool decision up to you but I think that we should favor code simplicity here generally and derisk any concerns with tests and observability.

deividaspetraitis replied:
For now I have removed to keep things simple as per you suggestion, we can always add this feature in the future if there is a need.

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -357,6 +363,64 @@ func (t *tokensUseCase) getChainScalingFactorMut(precision int) (osmomath.Dec, b
+
+// UpdateAssetsAtHeightInterval updates assets at specified height intervals by running fn.
+// Function fn will be launched as separate goroutine.
+func UpdateAssetsAtHeightInterval(height uint64, n int, fn LoadTokensFromChainRegistryFunc, us mvc.TokensUsecase, chainRegistryAssetsFileURL string) {
+	ExecuteOnTickInterval(height, n, func() {
+		StartAsRoutineIfNotRunning(func() {
+			fn(us, chainRegistryAssetsFileURL) // nolint
+		})
+	})
+}
```
Consider simplifying for readability.

Suggest doing something as simple as:
```go
if tick%uint64(interval) == 0 {
  go func() {
      // Handle asset list update
  }()
}
```

p0mvn commented on app/sidecar_query_server.go:
```diff
@@ -155,7 +155,27 @@ func NewSideCarQueryServer(appCodec codec.Codec, config domain.Config, logger lo
+			chainInfoUseCase,
+			appCodec,
+			quotePriceUpdateWorker,
+			[]func(height uint64){
+				func(height uint64) {
+					// do not block, run as a go routine
+					tokensUsecase.UpdateAssetsAtHeightInterval(
+						height,
+						200, // TODO config
+						tokensUsecase.LoadTokensFromChainRegistry,
+						tokensUseCase,
+						config.ChainRegistryAssetsFileURL,
+					)
+					logger.Info("callback block processing", zap.Uint64("height", height))
+				},
+			},
```
Not super sure why we need this wiring.

Can we instead have ingest handler call this:
```
tokensUsecase.UpdateAssetsAtHeightInterval(height)
```

For all the config parameters, we can set them on tokens use case in constructor

deividaspetraitis replied:
Original idea I had was to have more generic callback mechanism, where `ingester` just calls any number of `callback` functions with signature of `func(height uint64)`, and in turn specifics of constructing actual wiring were moved to the program entry point. 

As per your suggestion I have updated to just call `tokensUsecase.UpdateAssetsAtHeightInterval(height)` in the ingest handler.

deividaspetraitis replied:
Original idea I had was to have more generic callback mechanism, where `ingester` just calls any number of `callback` functions with signature of `func(height uint64)`, and in turn specifics of constructing actual wiring were moved to the program entry point. 

As per your suggestion I have updated to just call `tokensUsecase.UpdateAssetsAtHeightInterval(height)` in the ingest handler.

p0mvn commented:
A nit about mocks:

Let's say we have an `mvc.TokensUsecase` with 10 methods and `mvc.IngestUsecase` depends on the `mvc.TokensUsecase` but used only 2 methods.

Then, instead of creating a mock with 10 methods, we can make `mvc.IngestUsecase` depend on a separate interface with only 2 methods that `mvc.TokensUsecase` also implements.

This would allow us to customize the mock for this specific need without too much boilerplate.

Please let me know if anything doesn't make sense

cryptomatictrader commented:
@deividaspetraitis , I have a question about the implementation, does it update the entire asset list on every frequency, regardless whether there is any change? I wonder if it's worth adding the logic of storing the file size / checksum of the file in memory so that it won't even bother to touch the asset list at all?

deividaspetraitis commented:
> A nit about mocks:
> 
> Let's say we have an `mvc.TokensUsecase` with 10 methods and `mvc.IngestUsecase` depends on the `mvc.TokensUsecase` but used only 2 methods.
> 
> Then, instead of creating a mock with 10 methods, we can make `mvc.IngestUsecase` depend on a separate interface with only 2 methods that `mvc.TokensUsecase` also implements.
> 
> This would allow us to customize the mock for this specific need without too much boilerplate.
> 
> Please let me know if anything doesn't make sense

Just to confirm I get your idea right, continuing building on your idea, instead of:

```
func NewTokensUsecase(mvc.IngestUsecase) ... // mvc.IngestUsecase is 10 methods inferface, of which only 2 is used in constructed `TokensUsecase`
```

Have:

```
func NewTokensUsecase(mvc.Ingest2Methods) ... // mvc.Ingest2Methods is 2 methods inferface, of which all 2 are utilised in constructed `TokensUsecase`
```

Is that right?

deividaspetraitis commented:
> @deividaspetraitis , I have a question about the implementation, does it update the entire asset list on every frequency, regardless whether there is any change? I wonder if it's worth adding the logic of storing the file size / checksum of the file in memory so that it won't even bother to touch the asset list at all?

@cryptomatictrader, thank you for your time reviewing this PR and for feedback! You're totally right, current approach does the update regardless whether there is any change. Additionally, I think, we could try to determine whether we want to download assets data at all, there must be some HTTP header indicating whether contents has changed or not? In any case, it's probably question of priorities and I am happy to improve current implementation if you guys think it's worth spending the time on that at this point.

cryptomatictrader commented:
> > @deividaspetraitis , I have a question about the implementation, does it update the entire asset list on every frequency, regardless whether there is any change? I wonder if it's worth adding the logic of storing the file size / checksum of the file in memory so that it won't even bother to touch the asset list at all?
> 
> @cryptomatictrader, thank you for your time reviewing this PR and for feedback! You're totally right, current approach does the update regardless whether there is any change. Additionally, I think, we could try to determine whether we want to download assets data at all, there must be some HTTP header indicating whether contents has changed or not? In any case, it's probably question of priorities and I am happy to improve current implementation if you guys think it's worth spending the time on that at this point.

Thanks for the suggestion. The asset list is fetched from github site and I doubt they have some built-in header for us to perform some pre-flight check, plus the overhead of loading such file is minimal (just checked and it's around 150ms total response time). 

I think it's worth optimizing the update logic such that if there is no content change at all, we just skip the entire update process. We can do this by checking the total byte size of the response. What do you think?

deividaspetraitis commented:
> > > @deividaspetraitis , I have a question about the implementation, does it update the entire asset list on every frequency, regardless whether there is any change? I wonder if it's worth adding the logic of storing the file size / checksum of the file in memory so that it won't even bother to touch the asset list at all?
> 
> > 
> 
> > @cryptomatictrader, thank you for your time reviewing this PR and for feedback! You're totally right, current approach does the update regardless whether there is any change. Additionally, I think, we could try to determine whether we want to download assets data at all, there must be some HTTP header indicating whether contents has changed or not? In any case, it's probably question of priorities and I am happy to improve current implementation if you guys think it's worth spending the time on that at this point.
> 
> 
> 
> Thanks for the suggestion. The asset list is fetched from github site and I doubt they have some built-in header for us to perform some pre-flight check, plus the overhead of loading such file is minimal (just checked and it's around 150ms total response time). 
> 
> 
> 
> I think it's worth optimizing the update logic such that if there is no content change at all, we just skip the entire update process. We can do this by checking the total byte size of the response. What do you think? 

That definitely sounds like a plan, thank you for sharing ideas and testing possible approaches. Let me implement your suggested check once I am back working on this task.

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/v25.x' into 279

deividaspetraitis committed:
Update tokensUseCase to use sync.Map

With introduced auto-update asset list at regular intervals feature uncontrolled access may lead to panics

deividaspetraitis committed:
Requested change: simplifying implementation

deividaspetraitis committed:
Read height interval from config file

deividaspetraitis committed:
Cleanup, log LoadTokensFromChainRegistry errors

deividaspetraitis committed:
Requested change: when there is no content changes skip the update

deividaspetraitis committed:
Add entry to CHANGELOG.md

deividaspetraitis commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -349,26 +391,69 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
+
+// UpdateAssetsAtHeightInterval updates assets at configured height interval.
+// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
+func (t *tokensUseCase) UpdateAssetsAtHeightInterval(height uint64) {
```
Should we cover with tests this method? Any suggestions how we can do that?

p0mvn replied:
Yes. For the context of this method, we simply need to validate the following:
- `LoadTokensFromChainRegistry` is called at the expected height interval
- `LoadTokensFromChainRegistry` is not called at the expected height interval

Other invariants are to be validated by unit tests of the downstream methods.

We might have to rely on mocks to properly set this up.

deividaspetraitis replied:
Thanks, I have added test coverage for this method!

deividaspetraitis commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -349,26 +391,69 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
 
 // GetTokensFromChainRegistry fetches the tokens from the chain registry.
 // It returns a map of tokens by chain denom.
-func GetTokensFromChainRegistry(chainRegistryAssetsFileURL string) (map[string]domain.Token, error) {
```
Should we cover this function with tests?

p0mvn replied:
I think the ROI of covering this method by tests would be low assuming that we diligently unit test other methods.

It is hard to set up the test for this since we would have to mock out the HTTP request stuff. 

We also have higher-level functional tests for this component ensuring that this parses as intended. So I would advise to not spend too much time at this time.

deividaspetraitis commented:
Hey @p0mvn, @cryptomatictrader I have addressed your previous comments, can I ask please for one more round of reviews? I have also opened a few threads where your input would be appreciated. This PR is marked now as "ready for reviews", since based on your previous reviews I feel that we are getting close finishing it. :) Thanks!

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -349,26 +391,69 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
+
+// UpdateAssetsAtHeightInterval updates assets at configured height interval.
+// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
+func (t *tokensUseCase) UpdateAssetsAtHeightInterval(height uint64) {
```
nit: I think having this specified for the caller would be helpful
```suggestion
func (t *tokensUseCase) UpdateAssetsAtHeightIntervalAsync(height uint64) {
```

deividaspetraitis replied:
Thanks, it's updated including in all other relevant places.

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -349,26 +391,69 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
+	}
+}
+
+// LoadTokensFromChainRegistry loads tokens from the chain registry into TokensUsecase.
+// Tokens are loaded only if the contents of the chain registry file have changed.
+func (t *tokensUseCase) LoadTokensFromChainRegistry() error {
+	tokens, hash, err := GetTokensFromChainRegistry(t.chainRegistryAssetsFileURL)
+	if err != nil {
+		return err
+	}
+
+	if t.lastLoadTokensFromChainRegistryHash != hash {
+		t.LoadTokens(tokens)
+	}
+
+	return nil
 }
```
I think we should also unit-test this method. We should define `getTokensFromChainRegistry` as a field on `tokensUseCase`.

In core logic, we can wire up the actual implementation.

In the test logic, we can monkey-patch a mock.

Let me know if you'd like to have more pointers on how to get this working

deividaspetraitis replied:
Please see my comment on implementation: https://github.com/osmosis-labs/sqs/pull/323#discussion_r1665920048

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -411,8 +496,18 @@ func (t *tokensUseCase) RegisterPricingStrategy(source domain.PricingSourceType,
+		return false
+	}
+
+	v, ok := metaData.(domain.Token)
+	if !ok {
+		return false
+	}
```
Since this method starts growing in complexity, I suggest we add simple unit tests to ensure maximal test coverage while validating that the proper typing invariants are maintained.

deividaspetraitis replied:
I have covered this method with tests: https://github.com/osmosis-labs/sqs/pull/323/files#diff-393639508320686c9dbe3457d2678cac937a6294c260d8f30a59047d394f1193R548 

Note that I am unable cover with type assertion tests because  `tokensUseCase.coingeckoIds` is unexported and tests resides in different package meaning I am unable to load it with with invalid types ( for example `int` instead of `domain.Token` ). 

There must be some reasoning behind decision making tests to live in separate package, thus I just want to learn is current coverage enough or we should cover mentioned cases as well? 

IMHO is that tests living in the same packages would simplify testing in overall and allow to cover such edge cases.

p0mvn replied:
It is possible to mock out this test case by creating a test-specific method in `export_test.go`

```go
func (t *tokensUseCase) PutInvalidTypeTokenMetadata(key string, value any) {
	t.tokenMetadataByChainDenom.Store(key, value)
}
```

Note, that since this method exists in file `..._test.go` but `usecase` package, it would only be compiled for tests but not into the prod binary.

deividaspetraitis replied:
I just  tried  to go wtih your suggested approach and noticed that `tokensusecase.NewTokensUsecase` returns `mvc.TokensUsecase`, thus such method type is not available. If we want to go this route we can try return `TokensUseCase`  as contrete and exported type and then cast it in tests:

```
func NewTokensUsecase(tokenMetadataByChainDenom map[string]domain.Token, updateAssetsHeightInterval int, tokenRegistryLoader TokenRegistryLoader, logger log.Logger) mvc.TokensUsecase {
  ....
  return &TokensUseCase{...}
 ....
}
```

Tests:

```
usecase := tokensusecase.NewTokensUsecase(....)
usecase, ok := usecase.(tokensusecase.TokensUseCase)
```

Let me know in what appraoch you would favour.

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -446,9 +541,15 @@ func (t *tokensUseCase) IsValidPricingSource(pricingSource int) bool {
 		return "", fmt.Errorf("chain denom not found in chain registry")
 	}
+
+	v, ok := coingeckoId.(string)
+	if !ok {
+		return "", fmt.Errorf("coingecko id for chain denom (%v) is not of type string", chainDenom)
+	}
```
Ditto for covering with tests and all other similar instances.

p0mvn commented on ingest/usecase/ingest_usecase_test.go:
```diff
@@ -366,6 +370,48 @@ func (s *IngestUseCaseTestSuite) TestTransferDenomLiquidityMap() {
+		s.Require().NoError(err)
+	}
+
+	if want != got {
+		s.T().Fatalf("got %v, want %v", want, got)
+	}
```
I know this is probably code generated but we can do:
```go
s.Require().Equal(want, got)
```

p0mvn reviewed (commented):
Nice work!

Mostly had comments around adding tests but this is looking solid so far 👍

deividaspetraitis committed:
Requested change: Rename UpdateAssetsAtHeightInterval to UpdateAssetsAtHeightIntervalAsync

deividaspetraitis committed:
Requested change: Increase tests coverage

deividaspetraitis committed:
Requested changes: Cover IsValidChainDenom with tests

deividaspetraitis committed:
Requested change: update assertion

deividaspetraitis commented on tokens/usecase/token_registry_test.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/osmosis-labs/sqs/domain"
+)
+
+func TestFetchAndUpdateTokens(t *testing.T) {
```
This is very related with this comment: https://github.com/osmosis-labs/sqs/pull/323/files#r1665915262

I was not able to find elegant apprach how to write tests against `ChainRegistryHTTPFetcher` to check its internal state `initialHash` value without making test to live in the same packgae as t `ChainRegistryHTTPFetcher` implementation.  One of the options could possibly be to add public method to return `initialHash` but that would leak implementation details, with larger data structures that would result in too verbose API.

I understand that this approach is changing course of current approach and I am more than happy to align to it if we can came up with some ideas.

p0mvn commented on app/sidecar_query_server.go:
```diff
@@ -90,13 +90,22 @@ func NewSideCarQueryServer(appCodec codec.Codec, config domain.Config, logger lo
+			tokensusecase.GetTokensFromChainRegistry,
+		),
+		logger,
+	)
```
nit for readability
```go
	// Initialize chain registry HTTP fetcher
	chainRegistryHTTPFetcher := tokensusecase.NewChainRegistryHTTPFetcher(
		config.ChainRegistryAssetsFileURL,
		tokensusecase.GetTokensFromChainRegistry,
	)

	// Initialized tokens usecase
	tokensUseCase := tokensusecase.NewTokensUsecase(
		tokenMetadataByChainDenom,
		config.UpdateAssetsHeightInterval,
		chainRegistryHTTPFetcher,
		logger,
	)
```

p0mvn commented on tokens/usecase/token_registry.go:
```diff
@@ -0,0 +1,94 @@
+	}
+}
+
+// FetchAndUpdateTokensFunc is a FetchAndUpdateTokens function signature.
+type FetchAndUpdateTokensFunc func(loadTokens func(tokens map[string]domain.Token)) error
+
```
afaict this is unused

```suggestion
```

p0mvn commented on tokens/usecase/token_registry.go:
```diff
@@ -0,0 +1,94 @@
+
+// FetchAndUpdateTokens fetches tokens from the chain registry and loads by calling loadTokens  function.
+// In case there were no changes since last fetch, it does not call loadTokens.
+func (f *ChainRegistryHTTPFetcher) FetchAndUpdateTokens(loadTokens LoadTokensFunc) error {
```
Could we instead propagate this in the constructor of `ChainRegistryHTTPFFetcher`?

deividaspetraitis replied:
That was initial idea, however that approach led to some challanges when I was trying to came up how I can test `LoadTokensFunc` and wire everything.

When we construct `ChainRegistryHTTPFFetcher` we would need to pass `tokensUsecase.LoadTokens` method, but when we are constructing we need to pass `tokensUsecase`:

```go
	tokensUseCase := tokensusecase.NewTokensUsecase(
		...
		chainRegistryHTTPFetcher,
		...
	)

	// Initialize chain registry HTTP fetcher
	chainRegistryHTTPFetcher := tokensusecase.NewChainRegistryHTTPFetcher(
		...
		tokensUseCase.LoadTokens
		...
	)
```

One possible approach could be to just add extra method on ` tokensusecase`, such as `tokensusecase.SetTokenRegistryLoader(TokenRegistryLoader)`, but I did not done that for several reasons:

1. Use case would more difficult to initialise ( more wiring )
2. Use case might not work as not expeted in case someone forgot to set TokenRegistryLoader
3. I would still need to initialise `tokenusecase` first to have a reference to `tokensUsecase.LoadTokens` method

Let me know if I should proceed this or any other way! :rocket:

p0mvn commented on ingest/usecase/ingest_usecase_test.go:
```diff
@@ -366,6 +370,48 @@ func (s *IngestUseCaseTestSuite) TestTransferDenomLiquidityMap() {
+
+	if want != got {
+		s.Require().Equal(want, got)
+	}
```
```suggestion
	s.Require().Equal(want, got)
```

p0mvn commented on ingest/usecase/ingest_usecase_test.go:
```diff
@@ -366,6 +370,48 @@ func (s *IngestUseCaseTestSuite) TestTransferDenomLiquidityMap() {
+
+	want := rand.Intn(100)
+	for i := 0; i < want; i++ {
+		err = ingester.ProcessBlockData(context.TODO(), 1, nil, nil)
```
Please consider creating variables for the inputs so that it is clearer what it is that we are propagating as inputs

p0mvn commented on tokens/usecase/tokens_usecase_test.go:
```diff
@@ -539,5 +542,98 @@ func (s *TokensUseCaseTestSuite) TestGetMinPoolLiquidityCap() {
+
+// TestUpdateAssetsAtHeightIntervalAsync tests the async update of assets at height interval.
+func (s *TokensUseCaseTestSuite) TestUpdateAssetsAtHeightIntervalAsync() {
+	logger, err := log.NewLogger(false, "", "")
```
There is a `log.NoOpLogger` that can be helpful in tests

p0mvn commented on ingest/usecase/ingest_usecase_test.go:
```diff
@@ -366,6 +370,48 @@ func (s *IngestUseCaseTestSuite) TestTransferDenomLiquidityMap() {
+	)
+	s.Require().NoError(err)
+
+	want := rand.Intn(100)
```
Let's avoid pseudo-randomness in tests with a deterministic seed to prevent flakiness

p0mvn commented on tokens/usecase/tokens_usecase.go:
```diff
@@ -349,40 +387,29 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
+// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
+func (t *tokensUseCase) UpdateAssetsAtHeightIntervalAsync(height uint64) {
+	if height%uint64(t.updateAssetsHeightInterval) == 0 {
+		go func() {
+			if err := t.tokenLoader.FetchAndUpdateTokens(t.LoadTokens); err != nil {
+				t.logger.Error("error loading tokens from chain registry", zap.Error(err))
+			}
+		}()
 	}
-
-	return tokensByChainDenom, nil
 }
```
Consider simplifying to the following:
```go
// UpdateAssetsAtHeightIntervalAsync updates assets at configured height interval.
// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
func (t *tokensUseCase) UpdateAssetsAtHeightIntervalAsync(height uint64) {
	if height%uint64(t.updateAssetsHeightInterval) == 0 {
		go func() {
			if err := t.updateAssetsAtHeightIntervalSync(height); err != nil {
				// TODO: add prometheus error counter so that we can create an alert for observability
				t.logger.Error("error loading tokens from chain registry", zap.Error(err))
			}
		}()
	}
}

func (t *tokensUseCase) updateAssetsAtHeightIntervalSync(height uint64) error {
	if height%uint64(t.updateAssetsHeightInterval) == 0 {
		if err := t.tokenLoader.FetchAndUpdateTokens(t.LoadTokens); err != nil {
			return err
		}
	}
	return nil
}
```

Note that by having a separate sync version and returning an error, we improve the testability of the method, allowing for the ability to set up table-driven test cases and validate all branches.

Also, please note my TODO about the Prometheus metric

p0mvn replied:
Note that this ignores one of my previous comments about initializing `t.tokenLoader` with `t.LoadTokens` rather than propagating it for every call.

deividaspetraitis replied:
Totally agree, that simplies a lot of the logic in tests as well. Thanks for suggestion, implemened and pushed relevant changes!

p0mvn commented on tokens/usecase/tokens_usecase_test.go:
```diff
@@ -539,5 +542,98 @@ func (s *TokensUseCaseTestSuite) TestGetMinPoolLiquidityCap() {
+
+	tokensUsecase := tokensusecase.NewTokensUsecase(nil, updateAssetsHeightInterval, mock, logger)
+
+	for i := gen; i < gen+iterations+1; i++ {
+		tokensUsecase.UpdateAssetsAtHeightIntervalAsync(uint64(i))
+	}
```
Please see my other comment about separating a sync version of this so that we can set up a table-driven test case

p0mvn commented on tokens/usecase/tokens_usecase_test.go:
```diff
@@ -539,5 +542,98 @@ func (s *TokensUseCaseTestSuite) TestGetMinPoolLiquidityCap() {
+	logger, err := log.NewLogger(false, "", "")
+	s.Require().NoError(err)
+
+	updateAssetsHeightInterval, gen, iterations := rand.Intn(100), rand.Intn(8932122), rand.Intn(1000)
```
Ditto for using deterministic seeds as needed. Generally, I recommend opting for table-driven test cases , where possible, making the async version a simple propagator/wrapper to a heavily unit-tested synchronous method

p0mvn reviewed (commented):
Good work!

Getting close. Once the comments from the current review round are addressed, I expect to leave another cycle of comments focusing more on tests. Please let me know once ready.

Also, if you find many comments overwhelming, please feel free to open a fresh PR to reduce the comment clutter (but up to you 👍 )

cryptomatictrader commented on tokens/usecase/token_registry.go:
```diff
@@ -0,0 +1,94 @@
+
+// GetTokensFromChainRegistry fetches the tokens from the chain registry.
+// It returns a map of tokens by chain denom.
+func GetTokensFromChainRegistry(chainRegistryAssetsFileURL string) (map[string]domain.Token, string, error) {
```
Can we issue `response.Body.Close()` only once, and read the `response.Body` once ? Or there is a reason for that? I'm not sure what is the purpose of line 34 - line 39. Could you please explain? Thank you!

deividaspetraitis replied:
Oh, that's a good catch, thank you! I am not sure how I have ended up with this piece of code. I have updated response to be read once and then data to be reused for several operations. Please let me know if that makes sense?

cryptomatictrader commented on tokens/usecase/token_registry.go:
```diff
@@ -0,0 +1,94 @@
+	if err != nil {
+		return nil, "", err
+	}
+
```
Start of the code needs explanation.

cryptomatictrader commented on tokens/usecase/token_registry.go:
```diff
@@ -0,0 +1,94 @@
+	}
+	response.Body.Close() //  close old response body
+	response.Body = io.NopCloser(bytes.NewBuffer(data))
+
```
End of the code needs explanation

cryptomatictrader commented on tokens/usecase/token_registry.go:
```diff
@@ -0,0 +1,94 @@
+		return err
+	}
+
+	if f.lastFetchHash != hash {
```
👍 Nice optimization

cryptomatictrader reviewed (commented):
Nice work! Some minor issue / question to be addressed. Thank you!

deividaspetraitis committed:
Requested change: readability improvement

deividaspetraitis committed:
Requested changes: tests coverage

deividaspetraitis committed:
Requested change: Read response data once in GetTokensFromChainRegistry

deividaspetraitis committed:
Cleanup

deividaspetraitis commented:
I am closing this pull request in favour of new new one, since this one is starting to get hard to follow.
